processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -130,13 +156,18 @@ protected Integer addToFacts(
 
       // Last ditch sanity checks
       if (numEntries.get() >= maxRowCount
-          && facts.getPriorIndex(key) == TimeAndDims.EMPTY_ROW_INDEX
-          && !skipMaxRowsInMemoryCheck) {
-        throw new IndexSizeExceededException("Maximum number of rows [%d] reached", maxRowCount);
+          || maxBytesInMemory != -1 && sizeInBytes.get() >= maxBytesInMemory
+             && facts.getPriorIndex(key) == TimeAndDims.EMPTY_ROW_INDEX
+             && !skipMaxRowsInMemoryCheck) {
+        throw new IndexSizeExceededException("Maximum number of rows [%d] or max size in bytes [%d] reached", maxRowCount, maxBytesInMemory);
       }
       final int prev = facts.putIfAbsent(key, rowIndex);
       if (TimeAndDims.EMPTY_ROW_INDEX == prev) {
         numEntries.incrementAndGet();
+        if (maxBytesInMemory != -1) {
+          long estimatedRowSize = estimateRowSizeInBytes(key) + maxBytesPerRowForAggregators;
```

Comment:
The method `estimateRowSizeInBytes` should really be named `estimateKeySizeInBytes`. Or, it should keep the name `estimateRowSizeInBytes`, but in that case it should do the `+ maxBytesPerRowForAggregators` part. The point being that the row includes both the timeAndDims key, and the aggregators.

Reply:
Thinking about it a bit more, I think it makes sense to keep the method named `estimateRowSizeInBytes` and to add the `+ maxBytesPerRowForAggregators` into the method. That way, there's a clear entry point for the logic for estimating a row size, and people can look there if they want to improve it later.

Reply:
changed to second comment